src/cockpit/channels/filesystem.py

```diff
@@ -77,39 +79,38 @@ def do_open(self, options):
             self.close()
 
 
-class FsReadChannel(Channel):
+class FsReadChannel(GeneratorChannel):
     payload = 'fsread1'
 
-    def do_open(self, options):
-        self.ready()
-        try:
-            logger.debug('Opening file "%s" for reading', options['path'])
-            try:
-                with open(options['path'], 'rb') as filep:
-                    buf = os.stat(filep.fileno())
-                    tag = tag_from_stat(buf)
-                    if max_read_size := options.get('max_read_size'):
-                        if buf.st_size > max_read_size:
-                            raise ChannelError('too-large')
-
-                    data = filep.read()
-            except FileNotFoundError:
-                self.close(tag='-')
-                return
-            except PermissionError:
-                raise ChannelError('access-denied')
-            except OSError:
-                raise ChannelError('internal-error')
+    def do_yield_data(self, size_hint: int, options: Dict[str, object]) -> GeneratorChannel.DataGenerator:
+        binary = options.get('binary', False)
+        max_read_size = options.get('max_read_size')
+
+        logger.debug('Opening file "%s" for reading', options['path'])
 
-            if 'binary' not in options:
-                data = data.replace(b'\0', b'').decode('utf-8', errors='ignore').encode('utf-8')
+        try:
+            with open(options['path'], 'rb') as filep:
+                buf = os.stat(filep.fileno())
+                if max_read_size is not None and buf.st_size > max_read_size:
+                    raise ChannelError('too-large')
+
+                while True:
+                    data = filep.read1(size_hint)
+                    if data == b'':
+                        break
+                    logger.debug('  ...sending %d bytes', len(data))
+                    if not binary:
+                        data = data.replace(b'\0', b'').decode('utf-8', errors='ignore').encode('utf-8')
```

Comment:
.. Oh, I suppose *this* is the place you meant which swallows `\0`? That seems okay -- how would it loop infinitely here? it still consumes reading the file blocks, after all.